File: dsa/linkedlist/SinglyLinkedList.java
package linkedlist;

import java.util.Scanner;

public class SinglyLinkedList {
    static Scanner sc = new Scanner(System.in);
    int size = 0;
    Node head;

    class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    public static void main(String[] args) {
        SinglyLinkedList ob = new SinglyLinkedList();
        while (true) {
            System.out.print("Press:\n" +
                    "1 to Insert\n" +
                    "2 to Display\n" +
                    "3 to Delete\n" +
                    "4 to Delete All\n" +
                    "5 to exit()\n" +
                    "Enter your choice -> ");
            int choice = sc.nextInt();
            switch (choice) {
                case 1:
                    ob.insert();
                    break;
                case 2:
                    ob.display();
                    break;
                case 3:
                    ob.delete();
                    break;
                case 4:
                    ob.deleteAll();
                    break;
                case 5:
                    System.exit(0);
                    break;
                default:
                    System.out.println("Wrong Choice!");
            }
        }
    }

    public void insert() {
        while (true) {
            System.out.print("Press:\n" +
                    "1 to Insertion at Beginning\n" +
                    "2 to Insertion Inbetween\n" +
                    "3 to Insertion at Last\n" +
                    "4 Go Back\n" +
                    "Enter your choice -> ");
            int choice = sc.nextInt();
            switch (choice) {
                case 1:
                    insertBegin();
                    break;
                case 2:
                    insertInBtw();
                    break;
                case 3:
                    insertAtLast();
                    break;
                case 4:
                    return;
                default:
                    System.out.println("Wrong Input!");
            }
        }
    }

    public void delete() {
        // If list is Empty
        if (head == null) {
            System.out.println("List is Empty!");
            return;
        }
        System.out.print("Press:\n" +
                "1 to Deletion at Beginning\n" +
                "2 to Deletion Inbetween\n" +
                "3 to Deletion at Last\n" +
                "Enter your choice -> ");
        int choice = sc.nextInt();
        switch (choice) {
            case 1:
                delBegin();
                return;
            case 2:
                delInBtw();
                return;
            case 3:
                delAtLast();
                return;
            default:
                System.out.println("Wrong Input!");
        }
    }

    public void insertBegin() {
        System.out.print("Enter your Data -> ");
        int data = sc.nextInt();
        Node newNode = new Node(data);
        // If list is Empty
        if (head == null) {
            head = newNode;
            System.out.println("Data Inserted!");
            size++;
            return;
        }
        // newNode next will point to firstNode
        newNode.next = head;
        // head is set to newNode
        head = newNode;
        System.out.println("Data Inserted!");
        size++;
    }

    public void insertInBtw() {
        System.out.print("Enter the Index -> ");
        int i = sc.nextInt();
        // If index is 0
        if (i == 0) {
            System.out.println("Invalid Input! Index Starts from 1.");
            return;
        }
        // If index is 1
        if (i == 1) {
            insertBegin();
            return;
        }
        // If index is greater than the list size
        if (i > size) {
            System.out.println("Invalid Input! List Size is: " + size);
            return;
        }
        System.out.print("Enter your Data -> ");
        int data = sc.nextInt();
        Node newNode = new Node(data);
        Node currentNode = head;
        // traverse till index
        for (int j = 0; j < i - 2; j++)
            currentNode = currentNode.next;
        // this assign the next of currentNode i.e. address of next element to the next of newNode
        newNode.next = currentNode.next;
        // this assign the newNode to the next of currentNode
        // i.e. address of newNode is assigned in the next of currentNode
        currentNode.next = newNode;
        System.out.println("Data Inserted!");
        size++;
    }

    public void insertAtLast() {
        System.out.print("Enter your Data -> ");
        int data = sc.nextInt();
        Node newNode = new Node(data);
        // If list is Empty
        if (head == null) {
            head = newNode;
            System.out.println("Data Inserted!");
            size++;
            return;
        }
        Node currentNode = head;
        // traverse till last element
        while (currentNode.next != null)
            currentNode = currentNode.next;
        // lastNode next is set to newNode
        currentNode.next = newNode;
        System.out.println("Data Inserted!");
        size++;
    }

    public void delBegin() {
        // head from the firstNode is set to secNode, this removes the firstNode
        head = head.next;
        System.out.println("Successfully Deleted!");
        size--;
    }

    public void delInBtw() {
        // If list has only 1 element
        if (head.next == null) {
            head = null;
            System.out.println("List Successfully Cleared!");
            size = 0;
            return;
        }
        System.out.print("Enter the Index -> ");
        int i = sc.nextInt();
        // If index is 0
        if (i == 0) {
            System.out.println("Invalid Input! Index Starts from 1.");
            return;
        }
        if (i == 1) {
            delBegin();
            return;
        }
        // If index is equal to list size
        if (i == size) {
            delAtLast();
            return;
        }
        // If index is greater than the list size
        if (i > size) {
            System.out.println("Invalid Input! List Size is: " + size);
            return;
        }
        Node currentNode = head;
        // traverse till index
        for (int j = 0; j < i - 2; j++)
            currentNode = currentNode.next;
        // this assigns the address of 2nd next element to the currentNode,
        // to the next of currentNode, removing the mid element
        currentNode.next = currentNode.next.next;
        System.out.println("Successfully Deleted!");
        size--;
    }

    public void delAtLast() {
        // If list have one element
        if (head.next == null) {
            head = null;
            System.out.println("List Successfully Cleared!");
            size = 0;
            return;
        }
        Node secLastNode = head;
        Node lastNode = head.next;
        // traverse till last element
        while (lastNode.next != null) {
            lastNode = lastNode.next;
            secLastNode = secLastNode.next;
        }
        // secLastNode is set to null, this removes the address, removing the last element
        secLastNode.next = null;
        System.out.println("Successfully Deleted!");
        size--;
    }

    public void deleteAll() {
        System.out.println("List Successfully Cleared!");
        head = null;
        size = 0;
    }

    public void display() {
        // If list is Empty
        if (head == null) {
            System.out.println("List is Empty!");
            return;
        }
        Node currentNode = head;
        System.out.print("[");
        while (currentNode != null) {
            System.out.print(currentNode.data + ", ");
            currentNode = currentNode.next;
        }
        System.out.println("\b\b]" + "\nList Size -> " + size);
    }
}

File: dsa_with_python/linked_list/singly_linked_list.py
from node import Node


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    def is_empty(self) -> bool:
        return self.head is None

    def insert_begin(self, data):
        new_node = Node(data)
        if self.is_empty():
            self.head = new_node
            print("Data Inserted!")
            self.size += 1
            return
        new_node.next = self.head
        self.head = new_node
        print("Data Inserted!")
        self.size += 1

    def insert_at_pos(self, data, pos: int):
        if pos == 0:
            print("Invalid Input! Index Starts from 1.")
        elif pos == 1:
            self.insert_begin()
        elif pos > self.size:
            print(f"Invalid Input! List Size is: '{self.size}'")
        else:
            new_node = Node(data)
            current_node = self.head
            for i in range(0, pos - 2):
                current_node = current_node.next
            new_node.next = current_node.next
            current_node.next = new_node
            print("Data Inserted!")
            self.size += 1

    def insert_at_last(self, data):
        new_node = Node(data)
        if self.is_empty():
            self.head = new_node
            print("Data Inserted!")
            self.size += 1
        else:
            current_node = self.head
            while current_node.next is not None:
                current_node = current_node.next
            current_node.next = new_node
            self.size += 1

    def delete_at_begin(self):
        if self.is_empty():
            print("List is Empty!")
        elif self.head.next is None:
            self.head = None
            print("List Cleared Successfully!")
            self.size = 0
        else:
            print(f"Successfully Deleted '{self.head.data}' from beginning.")
            self.head = self.head.next
            self.size -= 1

    def delete_at_pos(self, pos: int):
        if self.is_empty():
            print("List is Empty!")
        elif self.head.next is None:
            self.head = None
            print("List Successfully Cleared!")
            self.size = 0
        elif pos == 0:
            print("Invalid Input! Index Starts from 1.")
        elif pos == 1:
            self.delete_at_begin()
        elif pos == self.size:
            self.delete_at_last()
        elif pos > self.size:
            print(f"Invalid Input! List Size is: '{self.size}'")
        else:
            current_node = self.head
            for i in range(0, pos - 2):
                current_node = current_node.next
            print(
                f"Successfully Deleted '{current_node.next.data}' from position '{pos}'."
            )
            current_node.next = current_node.next.next
            self.size -= 1

    def delete_at_last(self):
        if self.is_empty():
            print("List is Empty!")
        elif self.head.next is None:
            self.head = None
            print("List Successfully Cleared!")
            self.size = 0
        else:
            current_node = self.head
            while current_node.next.next is not None:
                current_node = current_node.next
            print(f"Successfully Deleted  '{current_node.next.data}' from last.")
            current_node.next = None
            self.size -= 1

    def delete_all(self):
        if self.is_empty():
            print("List is Empty!")
        else:
            print("List Successfully Cleared!")
            self.head = None
            self.size = 0

    def display(self):
        if self.is_empty():
            print("List is Empty!")
        else:
            current_node = self.head
            print("[", end="")
            while current_node is not None:
                print(current_node.data, end=", ")
                current_node = current_node.next
            print(f"\b\b]")
            print(f"List size is -> {self.size}")
